swap/src/asb/network.rs

```diff
@@ -138,4 +162,280 @@ pub mod behaviour {
+            match &mut self.registration_status {
+                RegistrationStatus::RegisterOnNextConnection => match self.connection_status {
+                    ConnectionStatus::Disconnected => {
+                        self.connection_status = ConnectionStatus::Dialling;
+
+                        return Poll::Ready(NetworkBehaviourAction::DialPeer {
+                            peer_id: self.rendezvous_peer_id,
+                            condition: DialPeerCondition::Disconnected,
+                        });
+                    }
+                    ConnectionStatus::Dialling => {}
+                    ConnectionStatus::Connected => {
+                        self.registration_status = RegistrationStatus::Pending;
+                        self.register();
+                    }
+                },
+                RegistrationStatus::Registered { re_register_in } => {
+                    if let Poll::Ready(()) = re_register_in.poll_unpin(cx) {
+                        match self.connection_status {
+                            ConnectionStatus::Connected => {
+                                self.registration_status = RegistrationStatus::Pending;
+                                self.register();
+                            }
+                            ConnectionStatus::Disconnected => {
+                                self.registration_status =
+                                    RegistrationStatus::RegisterOnNextConnection;
+
+                                return Poll::Ready(NetworkBehaviourAction::DialPeer {
+                                    peer_id: self.rendezvous_peer_id,
+                                    condition: DialPeerCondition::Disconnected,
+                                });
+                            }
+                            ConnectionStatus::Dialling => {}
+                        }
+                    }
+                }
+                RegistrationStatus::Pending => {}
+            }
+
+            let inner_poll = self.inner.poll(cx, params);
+
+            // reset the timer if we successfully registered
+            if let Poll::Ready(NetworkBehaviourAction::GenerateEvent(
+                libp2p::rendezvous::Event::Registered { ttl, .. },
+            )) = &inner_poll
+            {
+                let half_of_ttl = Duration::from_secs(*ttl) / 2;
+
+                self.registration_status = RegistrationStatus::Registered {
+                    re_register_in: Box::pin(tokio::time::sleep(half_of_ttl)),
+                };
+            }
+
+            inner_poll
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use crate::network::test::{new_swarm, SwarmExt};
+        use futures::StreamExt;
+        use libp2p::swarm::SwarmEvent;
+
+        #[tokio::test]
+        async fn given_no_initial_connection_when_constructed_asb_connects_and_registers_with_rendezvous_node(
```

Comment:
Can we somehow test that we register only once and not bombard the rendezvous node? 
That was one problem of the previous impl using a separate behaviour and I don't see it covered with this test. Not sure how easily it can be added.

Reply:
I am pretty confident the current implementation doesn't do that but I can also see to add a test yes.